### DISCUSSION TITLE
How to add volar for vue support?

### DISCUSSION THREAD

**Discussion by sane6** [LazyVim/LazyVim#2697] [General]
I just started using vim and have trouble adding LSP for vue. I added vue to tree-sitter and tried adding it to lsp configuration.

``` lua
   -- add more treesitter parsers
  {
    "nvim-treesitter/nvim-treesitter",
    opts = {
      ensure_installed = {
        "bash",
        "html",
        "javascript",
        "json",
        "lua",
        "markdown",
        "markdown_inline",
        "python",
        "query",
        "regex",
        "tsx",
        "typescript",
        "vim",
        "yaml",
        "vue",
      },
    },
  },
```

```lua
  {
    "neovim/nvim-lspconfig",
    ---@class PluginLspOpts
    opts = {
      ---@type lspconfig.options
      servers = {
        -- pyright will be automatically installed with mason and loaded with lspconfig
        pyright = {},
        volar = {},
      },
      setup = {
        volar = function()
          require("lazyvim.util").lsp.on_attach(function(client, _)
            if client.name == "volar" then
              client.server_capabilities.documentFormattingProvider = false
            end
          end)
        end,
      },
    },
  },
```

I can even see that volar is attached to the buffer. But It's still not working.
```
 1 client(s) attached to this buffer: 
 
 Client: volar (id: 1, bufnr: [3])
 	filetypes:       vue
 	autostart:       true
 	root directory:  /home/sane/Projects/vite-project
 	cmd:             /home/sane/.local/share/nvim/mason/bin/vue-language-server --stdio
 
```

**Comment by alewis001** [comment]
I have a `lua/plugins/lsp.lua` which looks like:
```lua
return {
  {
    "williamboman/mason.nvim",
    opts = {
      ensure_installed = {
        "bash-language-server",
        "css-lsp",
        "eslint-lsp",
        "gopls",
        "helm-ls",
        "html-lsp",
        "json-lsp",
        "jdtls",
        "marksman",
        "phpactor",
        "typescript-language-server",
        "vue-language-server",
      },
    },
  },
  {
    "nvim-lspconfig",
    keys = {
      { "<leader>ce", "<cmd>EslintFixAll<cr>", desc = "ESLint Fix All" },
    },
    opts = {
      setup = {
        volar = function()
          require("lazyvim.util").lsp.on_attach(function(client, _)
            if client.name == "volar" then
              client.server_capabilities.documentFormattingProvider = false
            end
          end)
        end,
      },
    },
  },
}
```
The `"vue-language-server"` bit in the `ensure_installed` of mason adds Volar. The `client.server_capabilities.documentFormattingProvider = false` bit is disabling the Volar formatter because it was conflicting with the eslint that is set up in the project.

**Reply by alewis001** [reply]
Once you add the `"vue-language-server"` to the Mason list, do `<leader>cm` (or `:Mason`) to open Mason and check "vue-language-server" is installed or, use the Mason menu/UI that appears to prompt the installation.

**Reply by sane6** [reply]
@alewis001 Thanks for confirming I'm not crazy and that this should work. It turns out that vue-language-server needs different setup for versions >2. I found no sane guides on how to set it up and what is actually different that breaks our setup. Just downgrading vue-language-server to version `1.8.27` makes it work fine. [Here's a huge thread](https://github.com/vuejs/language-tools/issues/3925) that talks about this issue, I didn't have time to read everything but all the solutions that I tried to make version >2 work failed.

**Reply by alewis001** [reply]
Ah yes, sorry, I forgot to say that I had stuck with 1.8.27 due to V2 issues. I did find this guide on getting it to work... https://github.com/williamboman/mason-lspconfig.nvim/issues/371#issuecomment-1988153959

I've not tried it yet. I'm hoping that something can be done about the "hard-coded" paths in the config via mason but might still give it a try.

**Comment by novito** [comment]
@alewis001 thanks for sharing this! Are you able to have the typescript language server properly working in vue files? If so, could you share how you made that happen?

**Comment by alewis001** [comment]
See below for my more recent config since moving to Volar 2. Judging by `:LspInfo` Volar and tsserver are loading for vue files, and just tsserver for typescript files. I have no idea if it's the "right" way to do this but as far as I can tell it's working (well... I think it is 😄 , somebody might tell me I'm only seeing half the functionality as my config is not quite right 🤣 ). I'm still not confident in just "knowing" how to write plugin config files for Lazyvim or being able to read plugin config docs or read other people's files and translate them into to Lazyvim confidently. 

The whole Volar2/tsserver back and forth has left me totally confused. Especially when you add in mason, mason-lspconfig and nvim-lspconfig. 

I think @folke has added/updated the vue extra which is now using Volar 2 judging by https://www.lazyvim.org/extras/lang/vue. As you can see, what is in the extra has some differences to what I have. I'd always take what @folke produces over anything I have so, I might just switch to the extra instead or, at least try to make my config closer to what is in the extra.

```lua
{
  {
    "williamboman/mason.nvim",
    opts = {
      -- Don't add jdtls (Java) to this as that is installed and managed by the nvim-java plugin
      -- These are the ones that are not supported in mason-lspconfig.
      ensure_installed = {
        "markdownlint",
      },
    },
  },
  {
    "williamboman/mason-lspconfig.nvim",
    opts = {
      -- Don't add jdtls (Java) to this as that is installed and managed by the nvim-java plugin
      ensure_installed = {
        "bashls",
        "cssls",
        "eslint",
        "gopls",
        "helm_ls",
        "html",
        "jsonls",
        "marksman",
        "phpactor",
        "tsserver",
        "volar",
      },
    },
    handlers = {
      -- default handler (optional)
      function(server_name)
        require("lspconfig")[server_name].setup({})
      end,

      -- language-server specific setup handlers.
      --
      -- This is no longer required but is an example of how to write a handler
      -- ["volar"] = function()
      --   require("lspconfig").volar.setup({
      --     filetypes = { "vue", "javascript", "typescript", "typescriptreact", "javascriptreact" },
      --     init_options = {
      --       vue = {
      --         hybridMode = false,
      --       },
      --       typescript = {
      --         tsdk = vim.fn.getcwd() .. "node_modules/typescript/lib",
      --       },
      --     },
      --   })
      -- end,
    },
  },
  {
    "nvim-lspconfig",
    keys = {
      { "<leader>ce", "<cmd>EslintFixAll<cr>", desc = "ESLint Fix All" },
    },
    opts = {
      setup = {
        tsserver = function(_, opts)
          local mason_registry = require("mason-registry")
          local vue_language_server_path = mason_registry.get_package("vue-language-server"):get_install_path()
            .. "/node_modules/@vue/language-server"

          opts.init_options = {
            plugins = {
              {
                name = "@vue/typescript-plugin",
                location = vue_language_server_path,
                languages = { "vue" },
              },
            },
          }
          opts.filetypes = { "typescript", "javascript", "javascriptreact", "typescriptreact", "vue" }
        end,
      },
    },
  },
}
```

**Comment by alewis001** [comment]
I've tried the using the vue extra and that's working fine so I think I'm going to switch to that.

**Reply by mena234** [reply]
Finally, a working solution, thanks.

**Reply by hsheikhali1** [reply]
sooo.. I wasn't able to get this working using the vue extra.. there is no auto-complete at all but the LSP is running

**Reply by Tjlastnumber** [reply]
You probably upgraded vue-language-server to version 3.0.0 or above. That version isn’t backward compatible. I’d recommend checking out the upgrade guide. I haven’t updated my config yet either. Alternatively, you can roll back to vue-language-server 2.2.8 using the command:
:MasonInstall vue-language-server@2.2.8

**Reply by hsheikhali1** [reply]
The thing is, I just used the lazy extra and didn't do anything specific :(

**Reply by Tjlastnumber** [reply]
LazyExtra is configured to use vue-language-server (aka Volar), but starting from version 3.0.0, the config is no longer compatible with the 2.x setup. Plus, Volar has been renamed to vue_ls, so the old LazyExtra setup doesn’t work anymore.

You’ve got two options here:
	1.	Wait for LazyVim to officially update LazyExtra to support the new vue_ls, or
	2.	Manually override the config yourself.

I spent a few hours trying to fix it myself but didn’t get it working, so I’m just going to wait for the official update.

In the meantime, you can check out the related issues on GitHub for more context. https://github.com/vuejs/language-tools/discussions/5456

**Reply by davlet61** [reply]
> You probably upgraded vue-language-server to version 3.0.0 or above. That version isn’t backward compatible. I’d recommend checking out the upgrade guide. I haven’t updated my config yet either. Alternatively, you can roll back to vue-language-server 2.2.8 using the command: :MasonInstall vue-language-server@2.2.8

Thanks a lot @Tjlastnumber , I was about to go crazy. Damn, I should be checking for breaking changes more often before upgrading ...

**Reply by hsheikhali1** [reply]
Yeah it's unfortunate that the vue community didn't consider this sort of breaking change but I understand it why.. anyway rolling back to 2.2.8 works :/

**Reply by Fate-ui** [reply]
> You probably upgraded vue-language-server to version 3.0.0 or above. That version isn’t backward compatible. I’d recommend checking out the upgrade guide. I haven’t updated my config yet either. Alternatively, you can roll back to vue-language-server 2.2.8 using the command: :MasonInstall vue-language-server@2.2.8

Thank you very much

**Reply by OliverWich** [reply]
Just went down chasing mason rabbit holes myself and can confirm, rolling back vue-language-server to 2.2.11 fixed the issue.

Lets just hope https://github.com/LazyVim/LazyVim/pull/6238 can get merged soon!

(Oh and thank you @Tjlastnumber for the insight of why this was happening!)
